src/status.cc

```diff
@@ -58,11 +91,96 @@ void StatusPrinter::BuildEdgeStarted(const Edge* edge,
     printer_.SetConsoleLocked(true);
 }
 
-void StatusPrinter::BuildEdgeFinished(Edge* edge, int64_t end_time_millis,
-                                      bool success, const string& output) {
+void StatusPrinter::RecalculateProgressPrediction() {
+  time_predicted_percentage_ = 0.0;
```

Comment:
Weak recommendation:

I think this kind of routine should be separated out into a function that can be given to the main loop as a parameter

Here you use the strategy of "If so many edges match what we saw last time with X ration", but other strategies might be better suited for a given environment. E.g. My environment would be better suited with a rolling-average calculation  that I'd parameterize in the ninjafile and/or cli args. and future estimates adjusted based on the current ratio instead of clamping